From f19405717ea23e7b83fb6e0ecaf3d13bf9a8e585 Mon Sep 17 00:00:00 2001
From: Jiang Xin <zhiyou.jx@alibaba-inc.com>
Date: Thu, 26 Dec 2019 16:12:52 +0800
Subject: [PATCH 4/7] refs-hook: set env GIT_REFS_TXN_NO_HOOK in gc and
 pack-refs

Command `git-gc` and `git-pack-ref` will not change references but only
do necessary maintainance. Set `GIT_REFS_TXN_NO_HOOK=1` to bypass refs-hook.

Signed-off-by: Jiang Xin <zhiyou.jx@alibaba-inc.com>
---
 builtin/gc.c        | 3 +++
 builtin/pack-refs.c | 2 ++
 2 files changed, 5 insertions(+)

diff --git a/builtin/gc.c b/builtin/gc.c
index b335cffa33..6f24b92df1 100644
--- a/builtin/gc.c
+++ b/builtin/gc.c
@@ -562,6 +562,9 @@ int cmd_gc(int argc, const char **argv, const char *prefix)
 		OPT_END()
 	};
 
+	/* Refs not changed in this command, so do not run hooks in refs transaction */
+	setenv("GIT_REFS_TXN_NO_HOOK", "1", 1);
+
 	if (argc == 2 && !strcmp(argv[1], "-h"))
 		usage_with_options(builtin_gc_usage, builtin_gc_options);
 
diff --git a/builtin/pack-refs.c b/builtin/pack-refs.c
index cfbd5c36c7..5e6d57c9f6 100644
--- a/builtin/pack-refs.c
+++ b/builtin/pack-refs.c
@@ -20,5 +20,7 @@ int cmd_pack_refs(int argc, const char **argv, const char *prefix)
 	git_config(git_default_config, NULL);
 	if (parse_options(argc, argv, prefix, opts, pack_refs_usage, 0))
 		usage_with_options(pack_refs_usage, opts);
+	/* Refs not changed in this command, so do not run hooks in refs transaction */
+	setenv("GIT_REFS_TXN_NO_HOOK", "1", 1);
 	return refs_pack_refs(get_main_ref_store(the_repository), flags);
 }
-- 
patchwork
